File: platform/platform-impl/src/com/intellij/diagnostic/DialogAppender.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.diagnostic;

import com.intellij.idea.AppMode;
import com.intellij.openapi.diagnostic.Attachment;
import com.intellij.openapi.diagnostic.ExceptionWithAttachments;
import com.intellij.openapi.diagnostic.IdeaLoggingEvent;
import com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments;
import com.intellij.util.ExceptionUtil;
import com.intellij.util.concurrency.AppExecutorUtil;
import org.jetbrains.annotations.ApiStatus;

import java.util.*;
import java.util.concurrent.ScheduledExecutorService;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.LogRecord;

@ApiStatus.Internal
public final class DialogAppender extends Handler {
  private static final int MAX_EARLY_LOGGING_EVENTS = 5;

  private static volatile boolean ourDelay;

  private int myEarlyEventCounter;
  private final Queue<IdeaLoggingEvent> myEarlyEvents = new ArrayDeque<>();
  private final ScheduledExecutorService myExecutor = AppExecutorUtil.createBoundedScheduledExecutorService("DialogAppender", 1);

  //TODO android update checker accesses project jdk, fix it and remove
  public static void delayPublishingForcibly() {
    ourDelay = true;
  }

  public static void stopForceDelaying() {
    ourDelay = false;
  }

  @Override
  public void publish(LogRecord event) {
    if (event.getLevel().intValue() < Level.SEVERE.intValue() || AppMode.isCommandLine()) {
      // the dialog appender doesn't deal with non-critical errors
      // also, it makes no sense when there is no frame to show an error icon
      return;
    }

    IdeaLoggingEvent ideaEvent;
    Object[] parameters = event.getParameters();
    if (parameters != null && parameters.length > 0 && parameters[0] instanceof IdeaLoggingEvent) {
      ideaEvent = (IdeaLoggingEvent)parameters[0];
    }
    else {
      Throwable thrown = event.getThrown();
      if (thrown == null) return;
      ideaEvent = extractLoggingEvent(event.getMessage(), thrown);
    }

    synchronized (this) {
      if (LoadingState.COMPONENTS_LOADED.isOccurred() && !ourDelay) {
        processEarlyEventsIfNeeded();
        queueAppend(ideaEvent);
      }
      else {
        myEarlyEventCounter++;
        if (myEarlyEvents.size() < MAX_EARLY_LOGGING_EVENTS) {
          myEarlyEvents.add(ideaEvent);
        }
      }
    }
  }

  private void processEarlyEventsIfNeeded() {
    if (myEarlyEventCounter == 0) return;
    IdeaLoggingEvent queued;
    while ((queued = myEarlyEvents.poll()) != null) {
      myEarlyEventCounter--;
      queueAppend(queued);
    }
    if (myEarlyEventCounter > 0) {
      queueAppend(new IdeaLoggingEvent(DiagnosticBundle.message("error.monitor.early.errors.skipped", myEarlyEventCounter), new Throwable()));
    }
  }

  private void queueAppend(IdeaLoggingEvent event) {
    if (DefaultIdeaErrorLogger.canHandle(event)) {
      myExecutor.execute(() -> DefaultIdeaErrorLogger.handle(event));
    }
  }

  private static IdeaLoggingEvent extractLoggingEvent(Object messageObject, Throwable throwable) {
    String message = null;
    List<ExceptionWithAttachments> withAttachments = ExceptionUtil.findCauseAndSuppressed(throwable, ExceptionWithAttachments.class);
    if (!withAttachments.isEmpty() && withAttachments.get(0) instanceof RuntimeExceptionWithAttachments) {
      message = ((RuntimeExceptionWithAttachments)withAttachments.get(0)).getUserMessage();
    }
    if (message == null && messageObject != null) {
      message = messageObject.toString();
    }
    if (withAttachments.isEmpty()) {
      return new IdeaLoggingEvent(message, throwable);
    }
    else {
      List<Attachment> list = new ArrayList<>();
      for (ExceptionWithAttachments e : withAttachments) {
        Collections.addAll(list, e.getAttachments());
      }
      return LogMessage.eventOf(throwable, message, list);
    }
  }

  @Override
  public void flush() { }

  @Override
  public void close() { }
}

File: platform/platform-impl/src/com/intellij/diagnostic/DialogAppender.kt
// Copyright 2000-2026 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.diagnostic

import com.intellij.featureStatistics.fusCollectors.LifecycleUsageTriggerCollector
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.ExceptionWithAttachments
import com.intellij.openapi.diagnostic.IdeaLogRecord
import com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments
import com.intellij.openapi.diagnostic.UnhandledExceptionKind
import com.intellij.util.ExceptionUtil
import com.intellij.util.io.pagecache.impl.Throttler
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.job
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import org.jetbrains.annotations.ApiStatus
import java.util.ArrayDeque
import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.atomic.AtomicBoolean
import java.util.logging.Handler
import java.util.logging.Level
import java.util.logging.LogRecord

@ApiStatus.Internal
class DialogAppender : Handler() {
  private val MAX_EARLY_LOGGING_EVENTS = 20

  @Suppress("RAW_SCOPE_CREATION") // DialogAppender is a process-wide root logger handler.
  private val coroutineScope = CoroutineScope(SupervisorJob() + DiagnosticDispatchers.Default + CoroutineName("DialogAppender"))

  private var earlyEventCounter = 0
  private val earlyEvents = ArrayDeque<Entry>()
  private var loggerBroken = AtomicBoolean(false)

  /** An error that goes to the message pool. [throwable] is the real cause. See IJPL-254578. */
  private class Entry(val message: String?, val throwable: Throwable, val unhandledExceptionKind: UnhandledExceptionKind)

  override fun publish(event: LogRecord) {
    if (event.level.intValue() < Level.SEVERE.intValue() || loggerBroken.get()) return

    val throwable = event.thrown ?: return
    // `JulLogger` drops the `UnhandledException` wrapper and puts the kind here. See IJPL-254578.
    val kind = (event as? IdeaLogRecord)?.unhandledExceptionKind ?: UnhandledExceptionKind.HANDLED
    val entry = Entry(event.message, throwable, kind)
    synchronized(this) {
      if (LoadingState.APP_READY.isOccurred) {
        processEarlyEventsIfNeeded()
        queueEvent(entry)
      }
      else {
        earlyEventCounter++
        if (earlyEvents.size < MAX_EARLY_LOGGING_EVENTS) {
          earlyEvents.add(entry)
        }
      }
    }
  }

  private fun processEarlyEventsIfNeeded() {
    if (earlyEventCounter == 0) return

    while (true) {
      val entry = earlyEvents.poll() ?: break
      earlyEventCounter--
      queueEvent(entry)
    }

    if (earlyEventCounter > 0) {
      LifecycleUsageTriggerCollector.onEarlyErrorsIgnored(earlyEventCounter)
      earlyEventCounter = 0
    }
  }

  private fun queueEvent(entry: Entry) {
    coroutineScope.launch {
      processEvent(entry)
    }
  }

  private val oomReportsThrottler = Throttler(100, SECONDS)

  private fun processEvent(entry: Entry) {
    try {
      val app = ApplicationManager.getApplication()
      if (app == null || app.isExitInProgress || app.isDisposed()) return

      val throwable = entry.throwable
      val oomErrorKind = DefaultIdeaErrorLogger.getOOMErrorKind(throwable)
      if (oomErrorKind != null) {
        val shouldNotify = synchronized(oomReportsThrottler) {
          oomReportsThrottler.isTimeForNextRun(System.nanoTime())
        }
        if (shouldNotify) {
          LowMemoryNotifier.showNotification(oomErrorKind, /*oomError: */true)
        }
      }
      else {
        val withAttachments = ExceptionUtil.causeAndSuppressed(throwable, ExceptionWithAttachments::class.java).toList()
        val message = withAttachments.asSequence().filterIsInstance<RuntimeExceptionWithAttachments>().firstOrNull()?.userMessage
                      ?: entry.message
        val attachments = withAttachments.asSequence().flatMap { it.attachments.asSequence() }.toList()
        // always add to MessagePool, dialog notification will decide if it shows or not in IdeMessagePanel
        MessagePool.getInstance().addErrorMessage(
          LogMessage(throwable, message, attachments, entry.unhandledExceptionKind)
        )
      }
    }
    catch (e: Throwable) {
      loggerBroken.set(true)
      throw e
    }
  }

  @ApiStatus.Internal
  suspend fun awaitPendingJobs() {
    val currentPendingJobs = synchronized(this) {
      if (LoadingState.APP_READY.isOccurred) {
        processEarlyEventsIfNeeded()
      }
      coroutineScope.coroutineContext.job.children.toList()
    }
    currentPendingJobs.joinAll()
  }

  override fun flush() { }

  override fun close() { }
}
